src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Extension.cs

```diff
@@ -63,6 +64,321 @@ internal override string ExtensionName
+
+            return pooledBuilder.ToStringAndFree();
+
+            static void appendType(TypeSymbol type, StringBuilder builder)
+            {
+                if (type is NamedTypeSymbol namedType)
+                {
+                    appendNamedType(type, builder, namedType);
+                }
+                else if (type is TypeParameterSymbol typeParameter)
+                {
+                    appendTypeParameter(typeParameter, builder, isDefinition: false);
+                }
+                else if (type is ArrayTypeSymbol array)
+                {
+                    appendArrayType(array, builder);
+                }
+                else if (type is PointerTypeSymbol pointer)
+                {
+                    appendType(pointer.PointedAtType, builder);
+                    builder.Append('*');
+                }
+                else if (type is FunctionPointerTypeSymbol functionPointer)
+                {
+                    appendFunctionPointerType(functionPointer, builder);
+                }
+                else
+                {
+                    throw ExceptionUtilities.UnexpectedValue(type);
+                }
+            }
+
+            static void appendNamedType(TypeSymbol type, StringBuilder builder, NamedTypeSymbol namedType)
+            {
+                Debug.Assert(type.CustomModifierCount() == 0);
+
+                if (namedType.SpecialType == SpecialType.System_Void)
+                {
+                    builder.Append("void");
+                    return;
+                }
+
+                if (namedType.IsReferenceType)
+                {
+                    builder.Append("class ");
+                }
+                else if (namedType.IsValueType)
+                {
+                    builder.Append("valuetype ");
+                }
+
+                appendNamespace(namedType.ContainingNamespace, builder);
+                appendContainingType(namedType, builder);
+                builder.Append(namedType.Name);
+                appendArity(namedType, builder);
+                appendTypeArguments(namedType, builder);
+            }
+
+            static void appendNamespace(NamespaceSymbol ns, StringBuilder builder)
+            {
+                if (ns is not null && !ns.IsGlobalNamespace)
+                {
+                    appendNamespace(ns.ContainingNamespace, builder);
+                    builder.Append(ns.Name);
+                    builder.Append('.');
+                }
+            }
+
+            static void appendContainingType(NamedTypeSymbol namedType, StringBuilder builder)
+            {
+                // Note: using slash for nested type to match CIL: ECMA-335 I.10.7.2
+                if (namedType.ContainingType is { } containingType)
+                {
+                    appendContainingType(containingType, builder);
+                    builder.Append(containingType.Name);
+                    appendArity(containingType, builder);
```

Comment:
We probably shouldn't include arity this way. We should simply use metadata name. It may or may not include arity, but CLR uses only that name to match types